Simple utility continuum baseline checker

brechmos-stsci opened the pull request:
This PR will create a simple spectrum checker to see if the flux baseline is near zero as this is required for several of the other methods (e.g., find_lines which assumes the flux data is continuum subtracted).

Fixes https://github.com/astropy/specutils/issues/535

eteq commented on specutils/analysis/flux.py:
```diff
@@ -116,3 +118,57 @@ def _compute_equivalent_width(spectrum, continuum=1, regions=None):
+
+def is_continuum_near_zero(spectrum, eps=0.01):
+    """
+    Determine if the baseline continuum of the spectrum is near zero.
```
```suggestion
    Determine if the typical baseline of this spectrum is near zero. I.e., an
    estimate of whether or not the continuum has been subtracted.
```

eteq commented on specutils/analysis/flux.py:
```diff
@@ -116,3 +118,57 @@ def _compute_equivalent_width(spectrum, continuum=1, regions=None):
+
+    Notes
+    -----
+    The spectrum should be continuum subtracted before being passed to this
```
I think this note is not necessary, since the point is to check if this is true or not.

brechmos-stsci replied:
k

eteq commented on specutils/analysis/flux.py:
```diff
@@ -116,3 +118,57 @@ def _compute_equivalent_width(spectrum, continuum=1, regions=None):
+    """
+    Determine if the baseline continuum of the spectrum is near zero.
+
+    The value is scaled to match the sigma/standard deviation parameter of a
```
Is this paragraph maybe copied from something else?  I don't really see how it applies to this function

brechmos-stsci replied:
Yup... will remove.

eteq commented on specutils/analysis/flux.py:
```diff
@@ -116,3 +118,57 @@ def _compute_equivalent_width(spectrum, continuum=1, regions=None):
+        #return np.median(spectrum.flux) / median_absolute_deviation(spectrum.flux) < eps
+
+def warn_spectrum_continuum_subtracted(eps):
+    def real_decorator(function):
```
I think you want to use the `wraps`or `update_wrapper` functions here (https://docs.python.org/3.7/library/functools.html#functools.wraps) - those do things like properly  pass through the docstring in a way that makes sense.

brechmos-stsci replied:
k

eteq commented on specutils/analysis/flux.py:
```diff
@@ -116,3 +118,57 @@ def _compute_equivalent_width(spectrum, continuum=1, regions=None):
+    if hasattr(spectrum, 'uncertainty') and spectrum.uncertainty:
+        return np.median(spectrum.flux / spectrum.uncertainty.quantity) < eps
+    else:
+        raise Exception('Spectrum flux has units, eps does not, either include uncertainty or use units on eps')
```
I think this `else` case should instead use the  median absolute deviation (in the "scaled to be like sigma" form - https://docs.astropy.org/en/stable/api/astropy.stats.mad_std.html ) as the "uncertainty" estimate, rather than raising an error.  Then the decorator can default to using that case with an `eps` of .01 or something like that, and we can count on this function always running on a spectrum.

(This is what makes it a heuristic for continuum subtraction - not all spectra that have their continuum subtracted really will have this property, but it's a reasonable guess for many spectra.)

brechmos-stsci replied:
Right, but, this was my question in https://github.com/astropy/specutils/issues/535#issuecomment-542339293.

eteq replied:
Oops, sorry - to finish out that conversation here, I don't really understand your point there.  In the case of your example linked there, that looks to have done exactly what was expected: the fluxes are close together, and that closeness indicates the "uncertainty per pixel" is probably low, so therefore the fact that the average offset from zero is larger than the scatter is *exactly* the sign of "continuum not subtracted" that we are looking for.

brechmos-stsci commented:
> 1. I think the docstring needs to describe the flow of what happens given various values of eps and whether or not the uncertainty is present (but maybe you were waiting on a first pass?)

I was waiting for the first pass to be done before finishing up the documentation.

> 2. This function doesn't seem to be using the mask. Presumably the user does not want masked values to be included, so I'd say the thing to do is just s.flux[~s.mask] if a mask is present.

Right, I was waiting to make sure the rest was good, but I added it in now.

> 3. There needs to be some way to turn the decorator "off", which the warning message should explain. I suggest using the configuration item machinery (see http://docs.astropy.org/en/stable/config/#adding-new-configuration-items) with just a boolean config item like always_check_continuum that defaults to True but just causes the decorator to immediately return True if it's set to False. Then you tell the user to use the with version of the config item machinery to supress the check.

I added this in, but you or someone will have to check it to make sure it is implemented correctly .I looked at astropy to see how it was used there, and mimiced it.

> 4. Once the modification for the "no uncertainty and no unit" case is in, we should test this on some realistic spectrum. Maybe use the spectrum from the "getting started" example? The goal is to check whether or not the three cases actually yield sensible answers, and also do some timing tests relative to other analysis functions like a line flux measurement.

Sure.

eteq commented on specutils/__init__.py:
```diff
@@ -34,3 +35,16 @@ class UnsupportedPythonError(Exception):
+    Configuration parameters for specutils.
+    """
+
+    always_check_continuum = _config.ConfigItem(
```
Instead of "always", perhaps `do_continuum_function_check` or something like that?  Only because "always" could be taken as "for all functions" and it's really just "selected functions".

brechmos-stsci replied:
changed

eteq commented on specutils/analysis/flux.py:
```diff
@@ -116,3 +124,65 @@ def _compute_equivalent_width(spectrum, continuum=1, regions=None):
+    spectrum : `~specutils.spectra.spectrum1d.Spectrum1D`
+        The spectrum object over which the width will be calculated.
+
+    eps: float
```
`eps` -> `threshold`.  And also should say that it can be a flux-unit quantity or a float/dimensionless qunatity, and that determines if it's a flux threshold vs a "signal/noise" threshold

brechmos-stsci replied:
already done

eteq commented on specutils/analysis/flux.py:
```diff
@@ -116,3 +124,65 @@ def _compute_equivalent_width(spectrum, continuum=1, regions=None):
+    if uncertainty:
+        return np.median(flux / uncertainty.quantity) < eps
+    else:
+        warnings.warn('Spectrum does not appear to be continuum subtracted.', AstropyUserWarning)
```
I'm confused why this warning is here?

brechmos-stsci replied:
I think you saw an old PR.

eteq commented on specutils/analysis/flux.py:
```diff
@@ -116,3 +124,65 @@ def _compute_equivalent_width(spectrum, continuum=1, regions=None):
+            if check:
+                spectrum = args[0]
+                if not is_continuum_near_zero(spectrum, eps):
+                    warnings.warn('Spectrum does not appear to be continuum subtracted.',
```
This should also have some text that refers back to `always_check_continuum` and tells the user exactly what they need to do to turn it off (or links to a page in the docs that explains it).

brechmos-stsci replied:
I added a short message with a link back to the astropy configuration section.

eteq commented:
Oops, I think my review there was against an old version, but I didn't realize there was a new commit until I had already done the review.  So ignore anything that was fixed in 72446dc that's in my comments.

eteq commented on specutils/analysis/flux.py:
```diff
@@ -116,3 +124,78 @@ def _compute_equivalent_width(spectrum, continuum=1, regions=None):
+    def actual_decorator(function):
+        @wraps(function)
+        def wrapper(*args, **kwargs):
+            if check:
```
Hmm. I might be confused but I think this isn't quite right - I think you want to do 
```suggestion
            if check():
```
because a configuration item can be changed at runtime.

eteq replied:
Or alternatively,
```suggestion
            if check() if callable(check) else check:
```
which would let you pass in True/False in addition to a configuration item

brechmos-stsci replied:
Based on our conversation, this changed to an alternative form (that is correct :) ).

eteq commented on specutils/fitting/fitmodels.py:
```diff
@@ -97,6 +97,7 @@ def _consecutive(data, stepsize=1):
     return np.split(data, np.where(np.diff(data) != stepsize)[0]+1)
 
 
+@warn_continuum_below_threshold(threshold=0.01, check=True)
```
I think you don't want `check=True` here (or below), right?  Because then the `do_continuum_function_check` option gets ignored?

brechmos-stsci replied:
hah, good point

eteq commented on specutils/analysis/flux.py:
```diff
@@ -116,3 +124,83 @@ def _compute_equivalent_width(spectrum, continuum=1, regions=None):
+
+def is_continuum_below_threshold(spectrum, threshold=0.01):
+    """
+    Determine if the baseline of this spectrum is less than the threshold.
```
```suggestion
    Determine if the baseline of this spectrum is less than a threshold.
```

eteq commented on specutils/analysis/flux.py:
```diff
@@ -116,3 +124,83 @@ def _compute_equivalent_width(spectrum, continuum=1, regions=None):
+    Determine if the baseline of this spectrum is less than the threshold.
+    I.e., an estimate of whether or not the continuum has been subtracted.
+
+    If the threshold is an `~astropy.units.Quantity` then compare the median
```
```suggestion
    If ``threshold`` is an `~astropy.units.Quantity` with flux units, this 
    directly compares the median of the spectrum to the threshold.
```

eteq commented on specutils/analysis/flux.py:
```diff
@@ -116,3 +124,83 @@ def _compute_equivalent_width(spectrum, continuum=1, regions=None):
+    I.e., an estimate of whether or not the continuum has been subtracted.
+
+    If the threshold is an `~astropy.units.Quantity` then compare the median
+    of the flux to the threshold.
```
(goes with the above suggestion...)
```suggestion
```

eteq commented on specutils/analysis/flux.py:
```diff
@@ -116,3 +124,83 @@ def _compute_equivalent_width(spectrum, continuum=1, regions=None):
+    If the threshold is an `~astropy.units.Quantity` then compare the median
+    of the flux to the threshold.
+
+    If the threshold is a float then the spectrum's uncertainty will be
```
```suggestion
    If the threshold is a float or dimensionless quantity then the spectrum's uncertainty will be
```

eteq commented on specutils/analysis/flux.py:
```diff
@@ -116,3 +124,83 @@ def _compute_equivalent_width(spectrum, continuum=1, regions=None):
+    threshold is compared to the median of the flux divided by the
+    uncertainty.  If the uncertainty is not present then the threshold
+    is compared to the median of the flux divided by the
+    `~astropy.stats.median_absolute_deviation`.
```
If you take my suggestion below this should be
```suggestion
    `~astropy.stats.mad_std`.
```

eteq commented on specutils/analysis/flux.py:
```diff
@@ -116,3 +124,83 @@ def _compute_equivalent_width(spectrum, continuum=1, regions=None):
+    if uncertainty:
+        return np.median(flux / uncertainty.quantity) < threshold
+    else:
+        return np.median(flux) / median_absolute_deviation(flux) < threshold
```
This should be `astropy.stats.mad_std` instead of `astropy.stats. median_absolute_deviation`.  The difference is just a scaling factor, but that's important for making it interpretable as "signal divided by a standard-deviation-like noise"

eteq commented on specutils/analysis/flux.py:
```diff
@@ -116,3 +124,83 @@ def _compute_equivalent_width(spectrum, continuum=1, regions=None):
+    # we are going to calculate based on the S/N if the uncertainty
+    # exists.
+    if uncertainty:
+        return np.median(flux / uncertainty.quantity) < threshold
```
I'm just realizing needs a check to ensure the uncertainty is `'std'`.  E.g.`` uncertainty.uncertainty_type != 'std'``, otherwise it's not predictable what kind of ucnertainty is being used.

In principle we should be able to convert everything to std-like, but since it's not obvious how to do that, perhaps it should check if it's a 'std'-style uncertainty, and if not it produces a warning that it doesn't know what to do with the uncertainty unless it's 'std', and is defaulting to the case like when there's no uncertainty?

eteq commented on specutils/analysis/flux.py:
```diff
@@ -116,3 +124,83 @@ def _compute_equivalent_width(spectrum, continuum=1, regions=None):
+    # If the threshold has units then the assumption is that we want
+    # to compare the median of the flux regardless of the
+    # existence of the uncertainty.
+    if hasattr(threshold, 'unit'):
```
I think we want to treat *dimensionless* quantities as though they are floats.  So something like:
```suggestion
    if hasattr(threshold, 'unit') and not threshold.unit == u.dimensionless_unscaled:
```
?
